src/S3/S3Transfer/AbstractMultipartUploader.php

```diff
@@ -0,0 +1,475 @@
+     */
+    public function __construct
+    (
+        S3ClientInterface $s3Client,
+        array $requestArgs,
+        array $config,
+        ?string $uploadId = null,
+        array $parts = [],
+        ?TransferProgressSnapshot $currentSnapshot = null,
+        ?TransferListenerNotifier $listenerNotifier = null,
+    ) {
+        $this->s3Client = $s3Client;
+        $this->requestArgs = $requestArgs;
+        $this->validateConfig($config);
+        $this->config = $config;
+        $this->uploadId = $uploadId;
+        $this->parts = $parts;
+        $this->currentSnapshot = $currentSnapshot;
+        $this->listenerNotifier = $listenerNotifier;
+    }
+
+    /**
+     * @param array $config
+     *
+     * @return void
+     */
+    protected function validateConfig(array &$config): void
+    {
+        if (!isset($config['target_part_size_bytes'])) {
+            $config['target_part_size_bytes'] = S3TransferManagerConfig::DEFAULT_TARGET_PART_SIZE_BYTES;
+        }
+
+        if (!isset($config['concurrency'])) {
+            $config['concurrency'] = S3TransferManagerConfig::DEFAULT_CONCURRENCY;
+        }
+
+        if (!isset($config['request_checksum_calculation'])) {
+            $config['request_checksum_calculation'] = 'when_supported';
+        }
+
+        $partSize = $config['target_part_size_bytes'];
+        if ($partSize < self::PART_MIN_SIZE || $partSize > self::PART_MAX_SIZE) {
+            throw new \InvalidArgumentException(
+                "Part size config must be between " . self::PART_MIN_SIZE
+                ." and " . self::PART_MAX_SIZE . " bytes "
+                ."but it is configured to $partSize"
+            );
+        }
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getUploadId(): ?string
+    {
+        return $this->uploadId;
+    }
+
+    /**
+     * @return array
+     */
+    public function getParts(): array
+    {
+        return $this->parts;
+    }
+
+    /**
+     * Get the current progress snapshot.
+     * @return TransferProgressSnapshot|null
+     */
+    public function getCurrentSnapshot(): ?TransferProgressSnapshot
+    {
+        return $this->currentSnapshot;
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    public function promise(): PromiseInterface
+    {
+        return Coroutine::of(function () {
+            try {
+                yield $this->createMultipartUpload();
+                yield $this->processMultipartOperation();
+                $result = yield $this->completeMultipartUpload();
+                yield Create::promiseFor($this->createResponse($result));
+            } catch (Throwable $e) {
+                $this->operationFailed($e);
+                yield Create::rejectionFor($e);
+            } finally {
+                $this->callOnCompletionCallbacks();
+            }
+        });
+    }
+
+    /**
+     * @return PromiseInterface
+     */
+    protected function createMultipartUpload(): PromiseInterface
```

Comment:
I saw there were method name changes in https://github.com/yenfryherrerafeliz/aws-sdk-php/pull/5.  Are they going to be added here?